lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs

```diff
@@ -164,25 +194,34 @@ mkTx
     -> (Address -> Maybe (k 'AddressK XPrv, Passphrase "encryption"))
     -> CoinSelection
     -> Either ErrMkTx (Tx, SealedTx)
-mkTx proxy (TxPayload certs mkExtraWits) timeToLive (rewardAcnt, pwdAcnt) keyFrom cs = do
+mkTx proxy pm (TxPayload certs mkExtraWits) timeToLive (rewardAcnt, pwdAcnt) keyFrom cs = do
     let withdrawals = mkWithdrawals
             proxy
             (toChimericAccountRaw . getRawKey . publicKey $ rewardAcnt)
             (withdrawal cs)
 
     let unsigned = mkUnsignedTx timeToLive cs withdrawals certs
 
-    addrWits <- fmap Set.fromList $ forM (CS.inputs cs) $ \(_, TxOut addr _) -> do
-        (k, pwd) <- lookupPrivateKey keyFrom addr
-        pure $ mkWitness unsigned (getRawKey k, pwd)
+    wits <- case (txWitnessTagFor @k) of
+        TxWitnessShelleyUTxO -> do
+            addrWits <- fmap Set.fromList $ forM (CS.inputs cs) $ \(_, TxOut addr _) -> do
+                (k, pwd) <- lookupPrivateKey keyFrom addr
+                pure $ mkShelleyWitness unsigned (getRawKey k, pwd)
+
+            let withdrawalsWits
+                    | Map.null withdrawals = Set.empty
+                    | otherwise = Set.singleton $
+                      mkShelleyWitness unsigned (getRawKey rewardAcnt, pwdAcnt)
 
-    let withdrawalsWits
-            | Map.null withdrawals = Set.empty
-            | otherwise = Set.singleton $
-                mkWitness unsigned (getRawKey rewardAcnt, pwdAcnt)
+            pure $ (SL.WitnessSet (addrWits <> withdrawalsWits) mempty mempty)
+                <> mkExtraWits unsigned
 
-    let wits = (SL.WitnessSet (addrWits <> withdrawalsWits) mempty mempty)
-            <> mkExtraWits unsigned
+        TxWitnessByronUTxO -> do
+            bootstrapWits <- fmap Set.fromList $ forM (CS.inputs cs) $ \(_, TxOut addr _) -> do
+                (k, pwd) <- lookupPrivateKey keyFrom addr
+                pure $ mkByronWitness unsigned pm (getRawKey k, pwd)
+            pure $ SL.WitnessSet mempty mempty bootstrapWits
+                <> mkExtraWits unsigned --- TO_DO - this is needed?
```

Comment:
this one is maybe not needed - I do not know at this moment

Reply:
Yes, if extra witnesses are needed for certificates or whatnot, they end up here

Reply:
yeah, but in byron transactions it should not happen to my best knowledge, right?